Additional "How To" docs

luise opened the pull request.

luise commented:
@ejj , here are the remaining workflow stories. Would be great if you can continue the review (this is the second part of the PR I split up earlier)

aegamesi commented on docs/_HowKeldaWorks.md:
```diff
@@ -1,24 +1,40 @@
 
-![Kelda Diagram](Kelda_Diagram.png)
+### Daemon
+The first command,`kelda daemon`, starts a long-running process -- "the daemon".
```
should have a space after the first comma

luise replied:
Nice catch!

ejj commented on docs/_HowKeldaWorks.md:
```diff
@@ -1,24 +1,40 @@
+The first command, `kelda daemon`, starts a long-running process -- "the
+daemon". The daemon is responsible for booting machines in the cloud,
+configuring the machines, managing them, and rebooting them if anything goes
+wrong. This means that if a machine dies while no daemon is connected to it
```
"is connected to it" is ever so slightly misleading.  I would just say "is running".  The daemon doesn't need to be connected to reboot a failed machine.

luise replied:
Hm, yeah, I see how that's not completely accurate. I said "connected" because I wanted to signal that the daemon has to be registered with the namespace -- you can't just run `kelda daemon`.
How about "while no daemon is running for its deployment"? Or "while no daemon is tracking it"?

ejj commented on docs/_HowKeldaWorks.md:
```diff
@@ -1,24 +1,40 @@
+configuring the machines, managing them, and rebooting them if anything goes
+wrong. This means that if a machine dies while no daemon is connected to it
+(e.g. because you closed the laptop that's running the daemon), the machine will
+not be restarted. However, as soon as the daemon is reconnected to the
```
"As soon as the daemon comes back online"

ejj commented on docs/_HowKeldaWorks.md:
```diff
@@ -1,24 +1,40 @@
+### Run
+To launch an application, call `kelda run` with a JavaScript blueprint -- in
+this example `my_app.js`. The `run` command passes the parsed blueprint to the
+daemon, and the daemon sets up the infrastructure described in the blueprint.
```
So the one thing that's missing here.  (Could be in a different PR, if you think that's best please write a card).  We aren't really capturing the subtlety around `kelda run`ing a spec that's already up and running.  In that case, the daemon will just notice that machines exist rather than booting it.

There's some broader UX questions here (the current flow is kinda icky).   Again, this PR shouldn't fix everything w.r.t. this problem. Whatever you think.

luise replied:
Hm, yeah, I'll make a card for it. I'd be happy to add it in this PR, but I think it would require mentioning namespaces also which aren't documented anywhere.  So I think we should just add it all in a separate PR.

ejj commented on docs/_HowKeldaWorks.md:
```diff
@@ -1,24 +1,40 @@
 
-Kelda runs applications using Docker containers. You can think of a container
+### Containers
+Kelda runs applications using [Docker](http://docker.com/) containers. As shown
```
"using Docker"  I don't think you need the containers bit.

luise replied:
I kind of like saying containers here because it then is clearly connected to the containers in the next sentence. Alternatively, I'd delete "containers" in the first sentence, and then say "with three Docker containers" in the following sentence. I'll do that for now.

ejj commented on docs/_HowKeldaWorks.md:
```diff
@@ -27,9 +43,18 @@ isolated from other containers running on the same machine (unlike processes).
+scenario where a worker machine disappears. In that scenario, the master machine
+would reboot the application containers from the failed worker on one of the
+healthy worker machines. It is possible to have multiple master machines to
+safeguard against a master machine failing. If there are multiple master
```
I would ditch the "if there are multiple masters" sentence.  That's an implementation details that I don't think is relevant.

ejj commented on docs/_HowTo.md:
````diff
@@ -148,3 +148,85 @@ and check you get the right response:
 ```
+
+
+## How to Run the Daemon in Production
````
Can we think of a less scary phrase than "in Production"?

luise replied:
Hmm, I might just say "How to run the daemon".

ejj commented on docs/_HowTo.md:
```diff
@@ -148,3 +148,85 @@ and check you get the right response:
+_We recommend reading about [the daemon](#daemon) before reading this section._
+
+The default way to run the daemon is to run it on a local machine like your
+laptop. While this is easy, and might work for you, there are two drawbacks to
```
"a couple of drawbacks"

ejj commented on docs/_HowTo.md:
```diff
@@ -148,3 +148,85 @@ and check you get the right response:
+laptop. While this is easy, and might work for you, there are two drawbacks to
+keep in mind:
+
+* **Robustness**. If any machine dies while your laptop is offline, the machine
```
The word "machine" is kind of kelda internals slang.  I think it might be clearer if we say "Cloud VM" the first use, and then "VM" in future users.  Perhaps you can think of something even better?

ejj replied:
"your laptop" -> "the daemon"

"your daemon" -> "the daemon"

ejj commented on docs/_HowTo.md:
```diff
@@ -148,3 +148,85 @@ and check you get the right response:
+* **Collaboration**. Currently, there are two issues with using a local daemon
+  if you share the deployment with others.
+  - **Multiple Daemons**. Each daemon assumes it's the only daemon attached to
+  its deployment, so if your teammates are also running local daemons for your
```
I don't like "your teammates"  assumes too much about the reader.  maybe something along the lines of: if you're sharing your deployment with others that also have daemons or something.

ejj commented on docs/_HowTo.md:
```diff
@@ -148,3 +148,85 @@ and check you get the right response:
+_We recommend reading about [the daemon](#daemon) before reading this section._
+
+The default way to run the daemon is to run it on a local machine like your
+laptop. While this is easy, and might work for you, there are two drawbacks to
```
Btw do we really need this drawback section?  Can we just say "If you're running a long term deployment, we don't recommend runng the daemon on your laptop. " and just leave it at that?  

My concern is this subsection is kind of technical and scary sounding, and the drawbacks of running the daemon on your laptop are somewhat obvious (i.e. it's not running all the time).

luise replied:
Hm, yeah. I don't think the "collaboration" section is obvious though. We don't mention the part about multiple daemons anywhere, and the part about TLS keys also isn't mentioned directly anywhere. Let me try to rephrase and shorten it (and leave out the robustness point).

ejj commented on docs/_HowTo.md:
```diff
@@ -148,3 +148,85 @@ and check you get the right response:
+
+### Solution: A Shared Daemon on a Separate VM
+To solve the above problems, we recommend running a single, shared daemon on a
+small VM on your preferred cloud provider. Even though the daemon is remote, you
```
"your preferred cloud provider" -> "in the cloud"

ejj replied:
And perhaps ditch the "even though" sentence.  It doesn't add much

luise replied:
I actually think it makes a difference that I can run `kelda show` from my local laptop and don't have to ssh into the machines with the daemon, so I think I'll leave the  sentence.

ejj commented on docs/_HowTo.md:
```diff
@@ -148,3 +148,85 @@ and check you get the right response:
+small VM on your preferred cloud provider. Even though the daemon is remote, you
+can still run all `kelda` commands from your local machine.
+
+#### Setting up the Remote Daemon
```
Sigh ... this is somewhat painful.

We have a card on improving this workflow right?

luise replied:
Yup!

ejj commented on docs/_HowTo.md:
````diff
@@ -148,3 +148,85 @@ and check you get the right response:
+    ```
+  **Note!** Some providers (e.g. DigitalOcean) don't give you a private IP.
+  If that's the case, try the public IP of the VM instead. If that doesn't work
+  either, try `0.0.0.0`.
````
Why not just always recommend 0.0.0.0?

luise replied:
Oh, I'd much rather do that. Didn't know if that was frowned upon.

ejj commented on docs/_HowTo.md:
```diff
@@ -148,3 +148,85 @@ and check you get the right response:
+  If that's the case, try the public IP of the VM instead. If that doesn't work
+  either, try `0.0.0.0`.
+
+4. **Allow SSH**. To allow you and your teammates to SSH in to the deployment,
```
I wonder if this is complicated enough, that we should just tell people to ssh into the VM and do things from there.  What do you think?

luise replied:
Hmm, yeah, I could see that too. Basically, the story would just be "create a VM, SSH in, install everything, and get the blueprints to this VM, and run all kelda commands from the VM".  It's not pretty, but it might be more straight forward than this.
I would actually be fine with doing that for now.

ejj commented on docs/_HowTo.md:
````diff
@@ -1,5 +1,49 @@
+    // Deploy the application.
+    app.deploy(inf);
+    ```
+If your website is replicated, follow the guide for running a [replicated,
````
"replicated" is jargony.  Can we simplify the language?

luise replied:
Good point. Maybe "hosted on multiple servers"?

ejj commented:
LGTM, just some minor suggestions.  Ping me when you're ready.

luise commented:
@ejj , addressed your comments. Changed the "How to run the daemon" workflow to simply be 1) create a VM in the cloud, 2) install Kelda etc. and 3) run all Kelda CLI commands from this machine.

ejj commented:
Sounds good.  Rebase and I"ll merge it.

luise commented:
@ejj  , rebased
